File: bot.js
var irc = require('irc');
var api = require('./services/lib/api');
var opts = require('nomnom')
    .option('config', {
        'abbr': 'c',
        'default': 'bot.json',
        'help': 'config file to use'
    }).parse()

zenircbot = {
    setup: function() {
        zenircbot.config = api.load_config(opts.config);
        zenircbot.pub = api.get_redis_client(zenircbot.config.redis);
        zenircbot.sub = api.get_redis_client(zenircbot.config.redis);

        zenircbot.unsetRedisKeys();
        var cfg = zenircbot.server_config_for(0);
        console.log('irc server: '+cfg.hostname+' nick: '+cfg.nick);
        zenircbot.irc = new irc.Client(cfg.hostname, cfg.nick, cfg);
        var bot = zenircbot.irc

        bot.addListener('connect', function() {
            zenircbot.pub.set('zenircbot:nick', bot.nick);
            zenircbot.pub.set('zenircbot:admin_spew_channels',
                              cfg.admin_spew_channels)
        });

        bot.addListener('ctcp', function(nick, to, text, type) {
            console.log('action: ' + nick + ' said ' + text + ' to ' + to);
            if (to == bot.nick) {
                to = nick;
            }
            var msg = {
                version: 1,
                type: 'privmsg_action',
                data: {
                    sender: nick,
                    channel: to,
                    message: text.replace(/^ACTION /, '')
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('message', function(nick, to, text, message) {
            console.log(nick + ' said ' + text + ' to ' + to);
            if (to == bot.nick) {
                to = nick;
            }
            var msg = {
                version: 1,
                type: 'privmsg',
                data: {
                    sender: nick,
                    channel: to,
                    message: text
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('nick', function(oldNick, newNick) {
            zenircbot.pub.get('zenircbot:nick', function(err, nick) {
                if (nick == oldNick) {
                    zenircbot.pub.set('zenircbot:nick', newNick);
                }
            });
        });

        bot.addListener('join', function(channel, nick, message) {
            console.log(nick + ' joined ' + channel);
            var msg = {
                version: 1,
                type: 'join',
                data: {
                    sender: nick,
                    channel: channel
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('part', function(channel, nick, reason, message) {
            console.log(nick + ' left ' + channel + ' because ' + reason);
            var msg = {
                version: 1,
                type: 'part',
                data: {
                    sender: nick,
                    channel: channel
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('quit', function(nick, reason, channels, message) {
            console.log(nick + ' quit');
            var msg = {
                version: 1,
                type: 'quit',
                data: {
                    sender: nick
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('topic', function(channel, topic, nick, message) {
            console.log(nick + ' changed the topic in ' + channel + ' to "' + topic + '"');
            var msg = {
                version: 1,
                type: 'topic',
                data: {
                    sender: nick,
                    channel: channel,
                    topic: topic
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('names', function(channel, nicks) {
            console.log('Names: '+channel);
            console.log(nicks);
            var msg = {
                version: 1,
                type: 'names',
                data: {
                    channel: channel,
                    nicks: nicks
                }
            };
            zenircbot.pub.publish('in', JSON.stringify(msg));
        });

        bot.addListener('error', function(message) {
            console.log(message);
        });

        var output_handlers = {
            1: zenircbot.output_version_1
        };

        zenircbot.sub.subscribe('out');
        zenircbot.sub.on('message', function(channel, message) {
            var msg = JSON.parse(message);
            output_handlers[msg.version](bot, msg);
        });
    },
    unsetRedisKeys: function(){
        zenircbot.pub.del('zenircbot:nick');
        zenircbot.pub.del('zenircbot:admin_spew_channels');
    },
    server_config_for: function(idx) {
        var cfg = zenircbot.config.servers[idx];
        /* server-generic options */
        cfg.debug = zenircbot.config.options.debug;
        cfg.floodProtection = zenircbot.config.options.floodProtection;
        cfg.showErrors = zenircbot.config.options.debug;
        cfg.stripColors = zenircbot.config.options.stripColors;
        return cfg;
    },
    output_version_1: function(bot, message) {
        switch (message.type) {
        case 'privmsg':
            console.log('  privmsg');
            bot.say(message.data.to, message.data.message);
            break;
        case 'privmsg_action':
            console.log('  privmsg_action');
            bot.say(message.data.to, '\u0001ACTION ' + message.data.message + '\u0001');
            break;
        case 'topic':
            console.log('  topic');
            bot.send('TOPIC', message.channel, message.topic);
            break;
        case 'raw':
            console.log('  raw');
            bot.send(message.command);
            break;
        }
    }
}

zenircbot.setup();

File: bot.py
import gevent
import json
import argparse

from irc import IRCBot, run_bot
from gevent import monkey
from services.lib.api import load_config, get_redis_client


monkey.patch_all()

parser = argparse.ArgumentParser(
    description='ZenIRCBot, a new and different bot',
)
parser.add_argument('-c', '--config', default='./bot.json',
                    help='The config to use. (default: %(default)s)')
opts = parser.parse_args()

config = load_config(opts.config)
pub = get_redis_client(config['redis'])


class RelayBot(IRCBot):

    def __init__(self, *args, **kwargs):
        super(RelayBot, self).__init__(*args, **kwargs)
        pub.set('zenircbot:nick', self.conn.nick)
        gevent.spawn(self.do_sub)

    def do_sub(self):
        sub = get_redis_client(config['redis'])
        self.pubsub = sub.pubsub()
        self.pubsub.subscribe('out')
        for msg in self.pubsub.listen():
            message = json.loads(msg['data'])
            print 'Got %s' % message
            if message['version'] == 1:
                if message['type'] == 'privmsg':
                    self.respond(message['data']['message'],
                                 channel=message['data']['to'])

    def do_privmsg(self, nick, message, channel):
        if not channel:
            channel = nick
        to_publish = json.dumps({
            'version': 1,
            'type': 'privmsg',
            'data': {
                'sender': nick,
                'channel': channel,
                'message': message,
                },
            })

        pub.publish('in', to_publish)
        print 'Sending to in %s' % to_publish

    def do_part(self, nick, command, channel):
        to_publish = json.dumps({
            'version': 1,
            'type': 'part',
            'data': {
                'sender': nick,
                'channel': channel,
            }
        })
        pub.publish('in', to_publish)
        print 'Sending to in %s' % to_publish

    def do_quit(self, command, nick, channel):
        to_publish = json.dumps({
            'version': 1,
            'type': 'quit',
            'data': {
                'sender': nick,
            }
        })
        pub.publish('in', to_publish)
        print 'Sending to in %s' % to_publish

    def do_nick(self, old_nick, command, new_nick):
        if pub.get('zenircbot:nick') == old_nick:
            pub.set('zenircbot:nick', new_nick)
        print 'nick change: %s -> %s' % (old_nick, new_nick)

    def command_patterns(self):
        return (
            ('/privmsg', self.do_privmsg),
            ('/part', self.do_part),
            ('/quit', self.do_quit),
            ('/nick', self.do_nick),
        )


run_bot(RelayBot, config['servers'][0]['hostname'],
        config['servers'][0]['port'], config['servers'][0]['nick'],
        config['servers'][0]['channels'])
